libraries/AP_ExternalAHRS/AP_ExternalAHRS.cpp

```diff
@@ -0,0 +1,560 @@
+        return false;
+    }
+    return true;
+}
+
+/*
+  get filter status. We don't know the meaning of the status bits yet,
+  so assume all OK if we have GPS lock
+ */
+void AP_ExternalAHRS::get_filter_status(nav_filter_status &status) const
+{
+    memset(&status, 0, sizeof(status));
+    if (last_pkt1 && last_pkt2) {
+        status.flags.initalized = 1;
+    }
+    if (healthy() && last_pkt2) {
+        status.flags.attitude = 1;
+        status.flags.vert_vel = 1;
+        status.flags.vert_pos = 1;
+
+        const struct VN_packet2 &pkt2 = *last_pkt2;
+        if (pkt2.GPS1Fix >= 3) {
+            status.flags.horiz_vel = 1;
+            status.flags.horiz_pos_rel = 1;
+            status.flags.horiz_pos_abs = 1;
+            status.flags.pred_horiz_pos_rel = 1;
+            status.flags.pred_horiz_pos_abs = 1;
+            status.flags.using_gps = 1;
+        }
+    }
+}
+
+Vector3f AP_ExternalAHRS::get_gyro(void)
+{
+    if (!last_pkt1) {
+        return Vector3f();
+    }
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    WITH_SEMAPHORE(sem);
+    return Vector3f(pkt1.gyro[0], pkt1.gyro[1], pkt1.gyro[2]);
+}
+
+Vector3f AP_ExternalAHRS::get_accel(void)
+{
+    if (!last_pkt1) {
+        return Vector3f();
+    }
+    const struct VN_packet1 &pkt1 = *last_pkt1;
+    WITH_SEMAPHORE(sem);
+    return Vector3f(pkt1.accel[0], pkt1.accel[1], pkt1.accel[2]);
+}
+
+// send an EKF_STATUS message to GCS
+void AP_ExternalAHRS::send_status_report(mavlink_channel_t chan) const
+{
+    if (!last_pkt1) {
+        return;
+    }
+    // prepare flags
+    uint16_t flags = 0;
+    nav_filter_status filterStatus;
+    get_filter_status(filterStatus);
+    if (filterStatus.flags.attitude) {
+        flags |= EKF_ATTITUDE;
+    }
+    if (filterStatus.flags.horiz_vel) {
+        flags |= EKF_VELOCITY_HORIZ;
+    }
+    if (filterStatus.flags.vert_vel) {
+        flags |= EKF_VELOCITY_VERT;
+    }
+    if (filterStatus.flags.horiz_pos_rel) {
+        flags |= EKF_POS_HORIZ_REL;
+    }
+    if (filterStatus.flags.horiz_pos_abs) {
+        flags |= EKF_POS_HORIZ_ABS;
+    }
+    if (filterStatus.flags.vert_pos) {
+        flags |= EKF_POS_VERT_ABS;
+    }
+    if (filterStatus.flags.terrain_alt) {
+        flags |= EKF_POS_VERT_AGL;
+    }
+    if (filterStatus.flags.const_pos_mode) {
+        flags |= EKF_CONST_POS_MODE;
+    }
+    if (filterStatus.flags.pred_horiz_pos_rel) {
+        flags |= EKF_PRED_POS_HORIZ_REL;
+    }
+    if (filterStatus.flags.pred_horiz_pos_abs) {
+        flags |= EKF_PRED_POS_HORIZ_ABS;
+    }
+    if (!filterStatus.flags.initalized) {
+        flags |= EKF_UNINITIALIZED;
+    }
+
+    // send message
+    const struct VN_packet1 &pkt1 = *(struct VN_packet1 *)last_pkt1;
+    const float vel_gate = 5;
+    const float pos_gate = 5;
+    const float hgt_gate = 5;
+    const float mag_var = 0;
+    mavlink_msg_ekf_status_report_send(chan, flags,
```

Comment:
I wonder if we should hang a wrapper somewhere which takes the `filterStatus` and these extra fields.  We'd share that between EK2, EK3 and now this.

Reply:
possibly, but they aren't quite the same